### DISCUSSION TITLE
개발 비하인드

### DISCUSSION THREAD

**Discussion by jiyolla** [Remian103/lolskin-price-tracker#82] [Show and tell]
## Contents
[Maintainers](https://github.com/Remian103/lolskin-price-tracker/discussions/82#discussioncomment-1462164)
[Contributing](https://github.com/Remian103/lolskin-price-tracker/discussions/82#discussioncomment-1462166)
[ESLint 적용과정 (in CRA, VSCode)](https://github.com/Remian103/lolskin-price-tracker/discussions/82#discussioncomment-1550410)
[Deploy Options on Backend](https://github.com/Remian103/lolskin-price-tracker/discussions/82#discussioncomment-1550468)
[Actual migration of API server from single EC2 to ECS
](https://github.com/Remian103/lolskin-price-tracker/discussions/82#discussioncomment-1556695)

**Comment by jiyolla** [comment]
## Maintainers
- FE: @Remian103 
- BE: @jiyolla

**Comment by jiyolla** [comment]
## Contributing
 - Use discussion for any idea, suggestion and proposal that is rather abstract.
 - Use issue for concrete stuff.

**Comment by Remian103** [comment]
# ESLint 적용과정 (in CRA, VSCode)

CRA로 프로젝트를 시작하게 되면 기본적으로 eslint가 설치되어 있고 package.json 내의 eslintConfig 필드로 설정 가능하다.
기본적으로 react-app과 react-app/jest가 적용되어 있는 것을 확인할 수 있었다.
\
목표는 [Airbnb Style Guide](https://github.com/airbnb/javascript)을 추가하고, vscode 내에서 auto formatting을 위한 설정을 추가


## eslint 설정

### package 설치

1. [eslint-config-airbnb](https://www.npmjs.com/package/eslint-config-airbnb) 설치
```
yarn add eslint-config-airbnb --dev
```
> 패키지 설치 방법을 보면 해당 패키지의 peer dependencies로 ```eslint, eslint-plugin-import, eslint-plugin-react, eslint-plugin-react-hooks, eslint-plugin-jsx-a11y```가 제시되어 있는데, CRA로 생성된 프로젝트에는 이들 패키지가 자동으로 포함되는 것으로 확인된다. 고로 굳이 설치할 필요는 없어 보인다. (버전 호환 문제 발생 가능성은 존재함)

2. TypeScript 지원을 위해 [eslint-config-airbnb-typescript](https://www.npmjs.com/package/eslint-config-airbnb-typescript) 설치
```
yarn add eslint-config-airbnb-typescript --dev
```
> 마찬가지로 패키지 설치법에 ```@typescript-eslint/eslint-plugin@^4.29.3, @typescript-eslint/parser@^4.29.3``` 을 설치할 것을 요구하는데 CRA 프로젝트에는 자동으로 포함되는 패키지인 것으로 확인된다. 따라서 버전 호환 문제가 발생하지 않는 이상 별도로 설치할 필요는 없을 듯 하다.

### eslintConfig 설정
package.json 내의 eslintConfig 필드 내에 작성
```json
"eslintConfig": {
    "extends": [
        "react-app",
        "react-app/jest",
        "airbnb",
        "airbnb/hooks",
        "airbnb-typescript"
    ],
    "rules": {
        "indent": [ "error", 4, { "SwitchCase": 1 } ],
        "react/jsx-indent": [ "error", 4 ],
        "react/jsx-indent-props": [ "error", 4 ],
        "@typescript-eslint/indent": [ "error", 4 ],
        "object-curly-newline": [ "error", { "multiline": true } ],
        "react-hooks/exhaustive-deps": "warn",
        "max-len": "off",
        "jsx-a11y/click-events-have-key-events": "off",
        "no-restricted-syntax": "off",
        "jsx-a11y/no-static-element-interactions": "warn",
        "jsx-a11y/no-noninteractive-element-interactions": "warn"
    },
    "parserOptions": {
        "project": "./tsconfig.json"
    }
}
```
> rule 필드는 취향에 맞게 커스텀하였다.

## auto formatting 설정

처음에는 eslint 규칙에 맞게 auto formatting을 하기 위해서는 prettier가 필수인 줄 알았다.
airbnb style guide 적용을 위해 찾아본 자료들에서는 eslint와 prettier를 같이 설정하는 경우가 대부분이었기 때문이다.
하지만 vscode의 eslint extension으로도 eslint 규칙에 맞게 auto formatting이 가능하다는 것을 알게 되었고 prettier와 eslint와의 차이점도 어느정도 알게 되었다.
이해한 바로는 eslint는 주로 javascript의 다양한 문법 스타일을 하나로 강제하기 위해 쓰이는 툴이고 prettier는 indentation과 같은 code styling을 주로 담당하는 듯 하다.
해당 프로젝트에서 사용하고자 하는 airbnb style guide에는 code style을 강제하는 rule도 상당히 많고, 본래 목적이었던 auto formatting도 vscode extension으로 해결 가능하기 때문에, 굳이 프로젝트에 prettier를 추가해야 할 필요성을 느끼지 못했다.

따라서 해당 프로젝트에서는 prettier를 설치하지 않고 vscode의 eslint extension을 설치하여 eslint rule에 따른 auto formatting을 적용하기로 하였다.

### VSCode extension 설치
[eslint](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint)
![image](https://user-images.githubusercontent.com/23255023/139198088-68aaadf4-bda7-45ce-be9f-2bdbb606e038.png)

### auto formatting 설정
vscode에서 setting.json에 아래 필드 추가
```json
    // Set the default
    "editor.formatOnSave": false,
    // Enable per-language
    "[javascript]": {
        "editor.formatOnSave": true
    },
    "editor.codeActionsOnSave": {
        // For ESLint
        "source.fixAll.eslint": true
    }
```
위 설정을 추가하고 나면 source 파일을 작성 후 저장하면 eslint rule 에 맞게 auto fomatting이 이루어진다.

**Comment by jiyolla** [comment]
# Deploy Options on Backend

### Containerization
Containerization is good for many reasons. Ease of maintenance, reproducibility, scalibility, blahblah, anyway it's good!
All options below assume containerization.

### HTTPS
HTTPS is also considered a must-have nowadays, and all options below are capable of setting up the HTTPS support.

### Scalability
God bless the age of scalability, let's make it scale!
AFAIK, to scale, you prepare more workers(or simply scale) and manage the workers through load balancing.

#### For the literal scaling, we have following options:
1. Docker Compose on a single AWS EC2.
  Actually, it doesn't scale, or we can say it has a very low scale upperbound because it works on a single machine.
  Docker Compose however does provide some basic health check and restart of services.
2. AWS ECS on predefined AWS EC2s.
  AWS ECS manages preregistered machines and scales services on this predefined cluster.
3. AWS ECS on AWS Fargate.
  AWS ECS manages scales services on self managed infrastructures(cluster).
4. AWS Lambda.
  Similar to AWS ECS on Fargate, but with more restriction(or abstractions in a good way) on 'what' to run on 'which' trigger.

#### For load balancing, we have following options:
1. Some static load balancing: basic Nginx as reverse proxy or AWS API Gateway(I'd think it has much usage but any way).
2. More dynamic ones: Ingress Controllers in Cluster or AWS Load balancer.
  For ingress controllers there are plenty of options including but not limited to Nginx ingress and Traefik.
  Basically they themselves reside as node in cluster like Docker Swarm, Kubernetes, or some proprietary like AWS ECS(or EKS), 
  and act as a gateway distributing the traffic to the nodes in the cluster.
  AWS Load Balancer would be simply stand alone Ingress Controller for AWS ECS(EKS).


## Conclusion
I would like to go with:
AWS ECS on EC2 with AWS Load Balancer.

On EC2 for there is no free tier AWS Fargate, and migration from EC2 to fargate seems not require much effort.

With AWS Load Balancer for it might be easier to setup than Traefik.

I'm not using Serverless because the limit on what to run just sound not good to me(though there are several ways to solve that). And probably if I later work on bigger service, serverless will lose its main pros(Mostly cost on low traffic. Ease of maintaining sound doable on ECS as well) over other choices.


### Database?
Database is often the bottleneck when scaling.
Since AWS RDS seems provide decent scalability on paid use, just stick to the free tier usage of AWS RDS currently.


### CI/CD?
It deserves a separate post.


## References
[A detailed example of FastAPI deployment on ECS Fargate](https://medium.com/adobetech/deploy-microservices-using-aws-ecs-fargate-and-api-gateway-1b5e71129338)
[Traefik on Kubernetes](https://medium.com/@geraldcroes/kubernetes-traefik-101-when-simplicity-matters-957eeede2cf8)
[Things to consider on deployment by Official FastAPI docs](https://fastapi.tiangolo.com/deployment/concepts/)
[API Gateway v.s. ALB](https://dashbird.io/blog/aws-api-gateway-vs-application-load-balancer/)
[A thorough investigation of Options on deploying FastAPI(asgi) on Serverless](https://medium.com/swlh/fastapi-microservice-serverless-deployment-41a6d21e5cb3)

**Comment by jiyolla** [comment]
# Actual migration of API server from single EC2 to ECS

## Manage SSL/TLS server certificate
Previously, we've managed the SSL/TLS server certificate by [certbot](https://certbot.eff.org/) on EC2 with Nginx.

Since we are going to have [Application Load Balancer(ALB, a layer 7 Elastic Load Balancer)](https://aws.amazon.com/elasticloadbalancing/) as our new facade on API Server, we need to install SSL/TLS certificate on it.

The easiest way is using [AWS Certificate Manager(ACM)](https://aws.amazon.com/certificate-manager/).
Add the requested DNS records to validate and acquire(and auto renew) the certificate.


## Create the ALB
Create a temporary target group to create the ALB.

Lately, we will replace the target group with the actual service in ECS cluster.

#### About HTTP protocol between load balancer and application server.
[Uvicorn](https://www.uvicorn.org/)(the ASGI server we are using) currently only supports HTTP/1.1. So if we stick to Uvicorn, we should have load balancer to talk to Uvicorn container with HTTP/1.1.

Since the latency between load balancer and the ECS containers are going to be very low, either HTTP/1.1 or HTTP/2 would make no big difference.
https://stackoverflow.com/questions/41637076/http2-with-node-js-behind-nginx-proxy

Nevertheless, here is [the benefits of implement HTTP/2 end-to-end answered by a HTTP/2 maintainer](https://stackoverflow.com/a/36057190/15909723).

[Hypercorn](https://pgjones.gitlab.io/hypercorn/) supports HTTP/2.


## Create the [Elastic Container Service(ECS)](https://aws.amazon.com/ecs/) Cluster, Task Definition, Service
[A beautiful explanation of ECS](https://www.freecodecamp.org/news/amazon-ecs-terms-and-architecture-807d8c4960fd/).

Since we are using ECS on EC2 instead of Fargate, there are two levels of scaling.
Scaling on the EC2 instances in a cluster and scaling on tasks in a service.

Cluster level scaling can be configured in auto scaling group panel.
Service level scaling can be configured in service settings.

Cluster creation is rather simple, and most options could be adjusted later.

Task definition is docker compose file in AWS's language. Unfortunately they are not compatible, but we can use [`container-transform`](https://github.com/micahhausler/container-transform) to translate either form into another.

In our case, we'd start with a task definition having a single Uvicorn container, for this task to be scaled out even in single machine(for maximal utilization), set dynamic port mapping. 

Talking of replication, there are multiple choices mentioned in [offical FastAPI docs](https://fastapi.tiangolo.com/deployment/server-workers/). It doesn't state the exact reason why controlling the replication at container level is a better practice(rather it says there no rules in written in stone about it). I guess replication control(scale) at container level is often hardware-aware thus more intelligent, while setting static multiple workers on uvicorn(or gunicorn) could probably cause thrashing issue. But we do need more investigation on this.

Lastly, create the service.
Attach the ALB we defined and set proper scaling policy on task scaling.
